core/drivers/crypto/hisilicon/hpre_montgomery.c

```diff
@@ -0,0 +1,510 @@
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x98, 0xA9
+};
+
+/* big endian */
+static const uint8_t g_x448_gx[] = {
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x05
+};
+
+static const struct hpre_mgm_curve g_curve_list[] = {
+	{
+		.key_bits = 256,
+		.p = g_x25519_p,
+		.a = g_x25519_a,
+		.x = g_x25519_gx,
+	}, {
+		.key_bits = 448,
+		.p = g_x448_p,
+		.a = g_x448_a,
+		.x = g_x448_gx,
+	}
+};
+
+static TEE_Result
+hpre_montgomery_alloc_keypair(struct montgomery_keypair *key,
+			      size_t size_bits)
+{
+	size_t key_size = BITS_TO_BYTES(size_bits);
+
+	if (!key || (size_bits != X25519_KEY_BITS &&
+		     size_bits != X448_KEY_BITS)) {
+		EMSG("Invalid input parameter");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	key->priv = calloc(1, key_size);
+	if (!key->priv)
+		goto priv_err;
+
+	key->pub = calloc(1, key_size);
+	if (!key->pub)
+		goto pub_err;
+
+	return TEE_SUCCESS;
+pub_err:
+	free(key->priv);
+	key->priv = NULL;
+priv_err:
+	EMSG("HPRE montgomery alloc key pair fail");
+
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+static TEE_Result gen_random_privkey(uint8_t *priv, size_t key_bits)
+{
+	size_t key_size = BITS_TO_BYTES(key_bits);
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!priv) {
+		EMSG("Privkey param is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = hw_get_random_bytes(priv, key_size);
+	if (ret) {
+		EMSG("Fail to fill privkey");
+		return TEE_ERROR_NO_DATA;
+	}
+
+	return ret;
+}
+
+static enum hisi_drv_status
+hpre_montgomery_params_alloc(struct hpre_montgomery_msg *msg)
+{
+	uint32_t size = HPRE_MONTGOMERY_TOTAL_BUF_SIZE(msg->key_bytes);
+
+	msg->key = calloc(1, size);
+	if (!msg->key) {
+		EMSG("Fail to alloc montgomery key buf");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->key_dma = virt_to_phys(msg->key);
+	msg->in = msg->key + HPRE_X_KEY_SIZE(msg->key_bytes);
+	msg->in_dma = msg->key_dma + HPRE_X_KEY_SIZE(msg->key_bytes);
+	msg->out = msg->in + msg->key_bytes;
+	msg->out_dma = msg->in_dma + msg->key_bytes;
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void hpre_montgomery_params_free(struct hpre_montgomery_msg *msg)
+{
+	if (msg->key) {
+		memzero_explicit(msg->key, HPRE_X_KEY_SIZE(msg->key_bytes));
+		free(msg->key);
+		msg->key = NULL;
+	}
+}
+
+static enum hisi_drv_status
+hpre_montgomery_params_pretreatment(struct hpre_montgomery_msg *msg)
+{
+	uint8_t *p = msg->key;
+	uint8_t *a = p + msg->key_bytes;
+	uint8_t *k = a + msg->key_bytes;
+	uint8_t *u = msg->in;
+	uint8_t *dst = k;
+	uint32_t bsize = msg->key_bytes;
+	uint32_t dsize = msg->curve_bytes;
```

Comment:
What's the difference between `dsize` and `psize`? They always hold the same value.

Reply:
psize is the size of public key. Its value is the same as curve_bytes.

Reply:
maybe I can just use dsize here